server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorDriver.java

```diff
@@ -451,11 +463,17 @@ public void moveSegmentOut(final String sequenceName, final List<SegmentIdentifi
       final Committer committer
   )
   {
-    final List<String> sequenceNames;
+    final List<SegmentIdentifier> theSegments;
     synchronized (activeSegments) {
-      sequenceNames = ImmutableList.copyOf(publishPendingSegments.keySet());
+      final List<String> sequenceNames = ImmutableList.copyOf(publishPendingSegments.keySet());
+      theSegments = sequenceNames.stream()
+                                 .map(publishPendingSegments::remove)
```

Comment:
Better to call `publishPendingSegments.clear()` separately than this side-effect-based code